account_payment_line/models/account_payment.py

```diff
@@ -0,0 +1,271 @@
+            if self.payment_type == "inbound":
+                domain.append(("move_type", "=", "in_refund"))
+        elif self.partner_type == "customer":
+            if self.payment_type == "outbound":
+                domain.append(("move_type", "=", "out_refund"))
+            if self.payment_type == "inbound":
+                domain.append(("move_type", "in", ("out_invoice", "out_receipt")))
+        return domain
+
+    def _filter_amls(self, amls):
+        return amls.filtered(
+            lambda x: x.partner_id.commercial_partner_id.id
+            == self.partner_id.commercial_partner_id.id
+            and x.amount_residual != 0
+            and x.account_id.account_type in ("asset_receivable", "liability_payable")
+        )
+
+    def _hook_create_new_line(self, invoice, aml, amount_to_apply):
+        line_model = self.env["account.payment.counterpart.line"]
+        self.ensure_one()
+        return line_model.create(
+            {
+                "payment_id": self.id,
+                "name": "/",
+                "move_id": invoice.id,
+                "aml_id": aml.id,
+                "account_id": aml.account_id.id,
+                "partner_id": self.partner_id.commercial_partner_id.id,
+                "amount": amount_to_apply,
+            }
+        )
+
+    def action_propose_payment_distribution(self):
+        move_model = self.env["account.move"]
+        for rec in self:
+            if self.is_internal_transfer:
+                continue
+            domain = self._get_moves_domain()
+            pending_invoices = move_model.search(domain, order="invoice_date_due ASC")
+            pending_amount = rec.amount
+            rec.line_payment_counterpart_ids.unlink()
+            for invoice in pending_invoices:
+                for aml in self._filter_amls(invoice.line_ids):
+                    amount_to_apply = 0
+                    amount_residual = rec.company_id.currency_id._convert(
+                        aml.amount_residual,
+                        rec.currency_id,
+                        rec.company_id,
+                        date=rec.date,
+                    )
+                    if pending_amount >= 0:
+                        amount_to_apply = min(abs(amount_residual), pending_amount)
+                        pending_amount -= abs(amount_residual)
+                    rec._hook_create_new_line(invoice, aml, amount_to_apply)
+
+    def action_delete_counterpart_lines(self):
+        if self.line_payment_counterpart_ids and self.state == "draft":
+            self.line_payment_counterpart_ids = [(5, 0, 0)]
+
+    def _prepare_move_line_default_vals(self, write_off_line_vals=False):
+        res = super(AccountPayment, self)._prepare_move_line_default_vals(
+            write_off_line_vals
+        )
+        write_off_amount_currency = (
+            write_off_line_vals and write_off_line_vals.get("amount", 0.0) or 0.0
```

Comment:
write_off_line_vals will be a list of dictionaries also, please handle that case.